Port "Lifting State Up/Passing Down Event Handlers" to use npm/vite (#132)
Ada-Developers-Academy/core-unit-3: lifting-state-up/passing-down-event-handlers.md

    emailData: 'sooah@dev.org',
    isPresentData: false,
  },
  {
    id: 3,
    nameData: 'Chrissy',
    emailData: 'chrissy@dev.org',
    isPresentData: true,
  }
];

function App() {
  const [studentData, setStudentData] = useState(kInitialStudentData);

  return (
    <main>
      <h1>Attendance</h1>
      <StudentList students={studentData}></StudentList>
    </main>
  );
}
```

<br />

Notice how this allows the logic of the `App` component to be more condensed. Our eyes no longer need to jump over the large array of student data to find the logic of the component. And imagine still how much more difficult it would be to read if there were *multiple* large arrays of data in the component! Much of the art to arranging code is to do so in a way that makes it easy to *ignore* the parts that we don't need to pay attention to at the moment.

<br />

On your own, try moving the `kInitialStudentData` array into a separate file, and importing it into `App`. If moving it to a JavaScript file, Don't forget to export it from the file where you define it! This may resemble the following:

```js
export const kInitialStudentData = [ /* array contents */ ];
```

<br />

> Assuming that definition were in a file under the `src` folder named `data/studentData.js`, we could import it into `App` like this:

Author: "were" is subjunctive (for hypothetical situations). "was" is for something that actually happened already.

| `(studentId)`                                       | This function accepts one parameter: the ID of the student whose presence to toggle. Because there is only a single parameter, the parentheses may be omitted if desired. |
| `const students = ;`                                | For this function, we create a temporary array, `students` to contain the updated student array. |
| `studentData.map(...)`                              | We map over `studentData`, applying the function supplied as an argument to each student record in the array. `map` creates a new array with the result of each iterated function call being used to populate its values. |
| `student => {...}`                                  | `student` is the name of the parameter to the function we supplied to `map`. Within the function, we reference each iterated item as `student`...                                                                                                            |
| `if (student.id === studentId) {`                   | If we find the student that `studentId` indicates we should update... |

> | `return { ...student, isPresentData: !student.isPresentData }` | Copy the student (with object spread notation) and update the copy's `isPresentData` to be the opposite value of what it had (toggle it). By returning this copy, it will replace the previous version of the data for this student record. The original student record in the original array of student records is unmodified. |

Reviewer: ```suggestion
| `return { ...student, isPresentData: !student.isPresentData }` | Copy the student (with object spread notation) and update the copy's `isPresentData` to be the opposite value of what it was before (toggle it). By returning this copy, it will replace the previous version of the data for this student record. The original student record in the original array of student records is unmodified. |
```

</details>

### !end-callout

<!-- available callout types: info, success, warning, danger, secondary, star  -->
### !callout-info

## Refactoring Opportunity: Using Function-Passing State Updates

We can update arrays using the function-passing style of calling state update functions. Recall that the function we pass to the updater will be called with the current value of the state (our students array) as its first parameter. Our function should return a new value (a new students array) that React will use to update the related piece of state.

<br />

Think about what changes we would need to make to `toggleStudentPresence`, then expand the section below to see a possible solution.

<br />

<details>
<summary>Click here to see a possible solution</summary>

```js
  const toggleStudentPresence = (studentId) => {
    setStudentData(students => {
      return students.map(student => {
        if (student.id === studentId) {
          return { ...student, isPresentData: !student.isPresentData };
        } else {
          return student;
        }
      });
    });
  };
```

<br />

> While the behavior of both this version and the previous version are the same at present. If we extended the application to work with data from a server, this version would be more robust. It would ensure that the data we are updating is the most recent version of the data, rather than the version that was in state when the function was defined, avoiding potential bugs.

Reviewer: ```suggestion
While the behavior of both this version and the previous version are the same at present if we extended the application to work with data from a server, this version would be more robust. It would ensure that the data we are updating is the most recent version of the data, rather than the version that was in state when the function was defined, avoiding potential bugs.
```

Author: ```suggestion
While the behavior of both this version and the previous version are the same at present, if we extended the application to work with data from a server, this version would be more robust. It would ensure that the data we are updating is the most recent version of the data, rather than the version that was in state when the function was defined, avoiding potential bugs.
```
